src/lando/main/scm/git.py

```diff
@@ -0,0 +1,298 @@
+
+    @property
+    def repo_is_initialized(self) -> bool:
+        """Determine whether the target repository is initialised"""
+        return Path(self.path).exists() and self._git_call("status", cwd=self.path)
+
+    @classmethod
+    def repo_is_supported(self, path: str) -> bool:
+        """Determine wether the target repository is supported by this concrete implementation"""
+        # This only tests a remote target without any local filesystem interaction, the CWD doesn't matter.
+        return self._git_call("ls-remote", path, cwd="/tmp")
+
+    def _git_initialize(self):
+        self.refresh_from_db()
+
+        if self.is_initialized:
+            raise
+
+        result = self._git_run("clone", self.pull_path, self.name)
+        if result.returncode == 0:
+            self.is_initialized = True
+            self.save()
+        else:
+            raise Exception(f"{result.returncode}: {result.stderr}")
+
+    def _git_pull(self):
+        self._git_run("pull", "--all", "--prune")
+
+    def _git_apply_patch(
+        self, diff: str, commit_description: str, commit_author: str, commit_date: str
+    ):
+        f_msg = tempfile.NamedTemporaryFile(encoding="utf-8", mode="w+")
+        f_diff = tempfile.NamedTemporaryFile(encoding="utf-8", mode="w+")
+        with f_msg, f_diff:
+            f_msg.write(commit_description)
+            f_msg.flush()
+            f_diff.write(diff)
+            f_diff.flush()
+
+            self._git_run("apply", f_diff.name)
+
+            self._git_run("add", "-A")
+            self._git_run(
+                "commit",
+                "--date",
+                commit_date,
+                "--author",
+                commit_author,
+                "--file",
+                f_msg.name,
+            )
+
+    def _git_last_commit_id(self) -> str:
+        return self._git_run("rev-parse", "HEAD", cwd=self.path)
+
+    def _git_run(self, *args, cwd: Optional[str] = None, must_succeed: bool = True):
+        """Run a git command and return full output.
+
+        Parameters:
+
+        args: list[str]
+            Arguments to git
+
+        cwd: str
+            Optional path to work in, default to self.path
+
+        Returns: CompletedProcess[str]
+
+            An object with (at least) the following attributes: args, returncode, stdout and stderr
+        """
+        path = cwd or self.path
+        command = ["git"] + list(args)
+        logger.debug("Running " + " ".join(command) + " in " + path)
+        result = subprocess.run(
+            command, cwd=path, capture_output=True, text=True, env=self._git_env()
+        )
+
+        if must_succeed and result.returncode:
+            raise SCMException(
+                f"Error running git command; {command=}, {path=}, {result.stderr}",
+                result.stdout,
+                result.stderr,
+            )
+        return result.stdout.strip()
+
+    @staticmethod
+    def _git_call(*args, cwd: str) -> bool:
```

Comment:
IMO we should only implement `git_run` and remove `git_call`, and always return the full underyling `CompletedProcess`. We can check the return code as needed in the callers if we need to assert success/failure. The `subprocess.run` API is just a better version of `subprocess.call`, and always having access to the `CompletedProcess` would give us the most flexibility in handling command output. Generally I dislike less-flexible wrappers around APIs that are already well designed.

Reply:
The problem is that `_git_run` relies on the instance existing, while `_git_call` doesn't (e.g., doesn't rely on `self.path`). It's used for early tests by the class (to see if a give `pull_path` is supported by git), and is supposed to be a lightweight call.

Now that I think more about it, maybe I can just remove the dependency on `self.path` from _git_run, and use that 🤔